### PR TITLE
Updates dcp tutorial with recent updates to api including save, load, and distributed state dict

### PR DESCRIPTION
**Pull request by LucasLLC**
## Description
DCP has recently updated it's API's, introducing semantics which more closley mirror torch's save/load API's. This update shows users how to use behavior now included in dcp.save/load, as well as some additional details for using the distributed.state_dict methods for managing FQN's and parallelism defaults.

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

### CODE REVIEW

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 1]
```diff
@@ -22,8 +22,11 @@ In this tutorial, we show how to use DCP APIs with a simple FSDP wrapped model.
 How DCP works
 --------------
 
-:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel.
-In addition, checkpointing automatically handles fully-qualified-name (FQN) mappings across models and optimizers, enabling load-time resharding across differing cluster topologies.
+:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel. Users can use this module to save on any number of ranks in parallel,
```
Comment (svekars) [inline recipes_source/distributed_checkpoint_recipe.rst:25]:
```suggestion
:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel. You can use this module to save on any number of ranks in parallel,
```

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 2]
```diff
@@ -22,8 +22,11 @@ In this tutorial, we show how to use DCP APIs with a simple FSDP wrapped model.
...
+and then re-shard across differing cluster topologies at load time.
+
+Addditionally through the use of modules in :func:`torch.distributed.checkpoint.state_dict`, DCP offers support for gracefully handling state dict generation and loading in distributed settings.
```
Comment (svekars) [inline recipes_source/distributed_checkpoint_recipe.rst:28]: should it be ``state_dict``?
```suggestion
Addditionally, through the use of modules in :func:`torch.distributed.checkpoint.state_dict`,
DCP offers support for gracefully handling ``state_dict`` generation and loading in distributed settings.
```

**Review by svekars — COMMENTED** [review]
Thanks!

**Comment by LucasLLC** [comment]
Thanks @svekars great suggestions.

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 3]
```diff
@@ -22,8 +22,12 @@ In this tutorial, we show how to use DCP APIs with a simple FSDP wrapped model.
...
 --------------
 
-:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel.
-In addition, checkpointing automatically handles fully-qualified-name (FQN) mappings across models and optimizers, enabling load-time resharding across differing cluster topologies.
+:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel. You can use this module to save on any number of ranks in parallel,
+and then re-shard across differing cluster topologies at load time.
+
+Addditionally, through the use of modules in :func:`torch.distributed.checkpoint.state_dict`,
+DCP offers support for gracefully handling ``state_dict`` generation and loading in distributed settings.
+This includes managing fully-qualified-name (FQN) mappings across models and optimizers, and setting default parameters for PyTorch provided parallelisms.
```
Comment (wz337) [inline recipes_source/distributed_checkpoint_recipe.rst:30]: Is `setting default parameters` referring to the optimizer initialization? Just want to make sure I understand.

**Review by wz337 — APPROVED** [review]
Overall LGTM! Thanks Lucas!

File: recipes_source/distributed_checkpoint_recipe.rst (hunk 3 above)
Comment (LucasLLC) [reply recipes_source/distributed_checkpoint_recipe.rst:30]: Optimizer initialization is a good point, but I was referencing setting the state dict type for FSDP, since users don't have to do this if they use distributed state dict

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 4]
```diff
@@ -22,8 +22,12 @@ In this tutorial, we show how to use DCP APIs with a simple FSDP wrapped model.
 How DCP works
 --------------
 
-:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel.
-In addition, checkpointing automatically handles fully-qualified-name (FQN) mappings across models and optimizers, enabling load-time resharding across differing cluster topologies.
+:func:`torch.distributed.checkpoint` enables saving and loading models from multiple ranks in parallel. You can use this module to save on any number of ranks in parallel,
+and then re-shard across differing cluster topologies at load time.
```
Comment (fegin) [inline recipes_source/distributed_checkpoint_recipe.rst:26]: Should we also add `and different parallelisms`?

**Review by fegin — COMMENTED** [review]
This update looks good to me but lacks of the introduction of the ability to use `save` and `load` for `Stateful` objects.

**Comment by svekars** [comment]
@LucasLLC  can you address @fegin comments or will you do it in a separate PR?

**Comment by svekars** [comment]
For any follow up changes, please submit a new PR.

**Comment by LucasLLC** [comment]
Will circle back to this with requested changes! Thanks @fegin , @svekars
